pecos/core/ann/quantizer_impl/x86.hpp

```diff
@@ -0,0 +1,227 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+#pragma once
+#if defined(__GNUC__)
+#define PORTABLE_ALIGN32 __attribute__((aligned(32)))
+#else
+#define PORTABLE_ALIGN32 __declspec(align(32))
+#endif
+
+#pragma once
+#ifndef NO_MANUAL_VECTORIZATION
+#ifdef __SSE__
+#define USE_SSE
+#ifdef __AVX__
+#define USE_AVX
+#endif
+#endif
+#endif
+
+#if defined(USE_AVX) || defined(USE_SSE)
+#ifdef _MSC_VER
+#include <intrin.h>
+#include <stdexcept>
+#else
+#include <x86intrin.h>
+#endif
+#endif
+
+#include <algorithm>
+#include <limits>
+#include <vector>
+#pragma once
+#include "common.hpp"
+#include "utils/clustering.hpp"
+
+namespace pecos {
+
+namespace ann {
+
+    struct ProductQuantizer4Bits : ProductQuantizer4BitsBase {
+
+        __attribute__((__target__("default")))
+        void pack_codebook_for_inference() {
+            pack_codebook_for_inference_default();
+        }
+
+        __attribute__((__target__("avx512f")))
+        void pack_codebook_for_inference() {
+            local_codebooks.resize(original_local_codebooks.size(), 0);
+            for (index_type i = 0; i < num_local_codebooks; i++) {
+                for (size_t j = 0; j < num_of_local_centroids; j++) {
+                    for (int k = 0; k < local_dimension; k++) {
+                        local_codebooks[i * num_of_local_centroids * local_dimension + k * num_of_local_centroids + j]
+                            = original_local_codebooks[i * num_of_local_centroids * local_dimension + j * local_dimension + k];
+                    }
+                }
+            }
+        }
+
+        __attribute__((__target__("default")))
+        void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            pad_parameters_default(max_degree, code_dimension);
+        }
+
+        __attribute__((__target__("avx512f"))) void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
+            //  512 / 128 == 4 local codebooks at a time. Thus, will need to be adjust the code_dimension to make it divisible
+            //  by 4. If it's not, we have to pad 0 to extended dimensions.
+            code_dimension = code_dimension % 4 == 0 ? code_dimension : (code_dimension / 4 + 1) * 4;
```

Comment:
what's the relation between `code_dimension` versus `local_dimension` ?

Reply:
Same as below, code_dimension is from the perspective of HNSWPQ4Bits and local_dimension is from the perspective of quantizer. I will not pad  local_dimension but code_dimension could be padded if to fit memory layout in GraphPQ4Bits. In sum, code_dimension could be different from local_dimension.